### ISSUE TITLE
Support account's execute

### ISSUE THREAD

**Issue by fracek** [apibara/starknet-react#102]
This is different from #6. The account interface exposes an `execute` method to invoke multiple external functions in one transaction.

The proposed interface is:

 ```ts
const { execute, loading, error } = useStarknetExecute()
```

Where `execute` is:

```ts
execute: ({ calls: CallArgs, metadata?: any }) => Promise<AddTransactionResponse | undefined>
type CallArgs = { args?: T, contract?: Contract, method?: string }
```

## Example

```ts
const { token } = useTokenContract()
const { other } = useOtherContract()

const { account } = useStarknet()
const { execute } = useStarknetExecute()

const [amount, setAmount] = useState(0)

// ...

execute({
  metadata: { message: 'Approve token and swap' },
  calls: [
    { contract: token, method: 'approve', args: amount && account ? [amount, account] : undefined },
    { contract: other, method: 'swap', args: amount && account ? [amount, account /*, more data here */] : undefined },
  ]
})
```

Like `invoke`, `execute` checks that all calls arguments are defined before making the call.

**Comment by fracek** [comment]
Since `execute` does not require changing any existing interface it can be added after v1.0.

What's your opininion on this interface @tarrencev  @pet3ris?

**Comment by Pet3ris** [comment]
Would be great to have this function, I'm already using multicall right now for Yagi.

Regarding naming, I wonder if having more aligned naming between the invoke and this would make sense, e.g., `execute1` and `execute` or `invoke1` and `invoke` or `invoke` and `invokeMany`.

**Comment by fracek** [comment]
I like `invokeMany` but we would be diverging from the standard name for the function which (I believe) is confusing.

**Comment by tarrencev** [comment]
I like the proposed interface for execute. I'm wondering if we should deprecate `invoke` api in favor of a single execute api. It seems that is the direction the `AccountInterface` has taken:

```
public abstract execute(
    transactions: Call | Call[],
    abis?: Abi[],
    transactionsDetail?: InvocationsDetails
  ): Promise<AddTransactionResponse>;
```

**Comment by fracek** [comment]
Looks like the `execute` method only accepts `RawCalldata`, I wonder if there's a reason for that? We also need to figure out how to add typesafety to the hook, my impression is that it's enough to add a `unknown[]` type parameter, where each `unknown` is the type of the input of each call.

/cc @janek26

**Comment by ponderingdemocritus** [comment]
It would be a nice feature if the calls were view exposed so within a UI you could easily show a 'tx cart' style element. Think like a ecomm cart.

For games specifically, it might make sense to seriously batch transactions together, as if they are updating similar state values (erc20 transfer eg) in 10 transactions, then it is effectively 1 storage update. Extend this to an 1155 style and the savings stack up massively for the player.

**Comment by ponderingdemocritus** [comment]
![Screenshot from 2022-06-01 16-45-43](https://user-images.githubusercontent.com/90423308/171344468-f17a83d7-3612-4999-8317-f5204bb8535a.png)

Pulled from StarkNet fees - This is a huge potential saving if we batch transactions.

**Comment by fracek** [comment]
> It would be a nice feature if the calls were view exposed so within a UI you could easily show a 'tx cart' style element. Think like a ecomm cart.
> 
> For games specifically, it might make sense to seriously batch transactions together, as if they are updating similar state values (erc20 transfer eg) in 10 transactions, then it is effectively 1 storage update. Extend this to an 1155 style and the savings stack up massively for the player.

I agree this looks cool. To make it happen we need to figure out a good API for execute, then we can think about that type of component built on top.

**Comment by fracek** [comment]
Fixed in 9fb19b277723e508ed73
